### PR TITLE
Adds Readable stream support when saving

### LINKED ISSUE #66: Add Streams API Support
Using the streams API is it possible to pipe data into a file handler e.g.

```ts
const fileHandle = await showSaveFilePicker();
const writableStream = await fileHandle.createWritable();
compressedStream.pipeTo(writableStream);
```

Could we add support for passing a stream to this library to more efficiently handle saving large files.
Obviously passing a stream in legacy mode would either need to reject or build a blob from the stream and save that.

Would this been considered within the scope of this library? I'm open to working on a PR

### PR DESCRIPTION
**Pull request by seanaye**
This PR allows passing a readable stream to the fileSave function instead of a blob. When supported this will pipe bytes directly into a writable stream, which reduces memory usage on large files (prevents copying). When not supported, the stream will simply be converted to a blob

closes #66

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -98,8 +98,8 @@ export type WellKnownDirectory =
  * @returns Optional file handle to save in place.
  */
 export function fileSave(
-  /** To-be-saved blob */
-  blob: Blob,
+  /** To-be-saved blob or stream&mimeType object*/
+  blobOrStream: Blob | { readable: ReadableStream; type: string },
```
Comment (tomayac) [inline index.d.ts:102]: Do we need the object? Could it just be `ReadableStream`, as the variable name suggests?

File: src/fs-access/file-save.mjs [hunk 2]
```diff
@@ -67,6 +67,17 @@ export default async (
       excludeAcceptAllOption: options[0].excludeAcceptAllOption || false,
     }));
   const writable = await handle.createWritable();
+  // use streaming on the blob if the browser supports it
+  if ('stream' in blobOrStream) {
```
Comment (tomayac) [inline src/fs-access/file-save.mjs:71]: I don't get this part. There's no `stream()` method on `ReadableStream` ([MDN](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream#methods)). I may be misunderstanding what you want to do here.

Comment (tomayac) [reply src/fs-access/file-save.mjs:71]: Oh, sorry. This is referring to [`Blob.stream()`](https://developer.mozilla.org/en-US/docs/Web/API/Blob/stream). Ignore my comment above.

**Review by tomayac — COMMENTED** [review]
I think this is headed in a great direction. Just one remark regarding the wrapping object. (Ignore the other comment that I have clarified myself.)

File: src/fs-access/file-save.mjs [hunk 3]
```diff
@@ -67,6 +67,17 @@ export default async (
       excludeAcceptAllOption: options[0].excludeAcceptAllOption || false,
     }));
   const writable = await handle.createWritable();
+  // use streaming on the blob if the browser supports it
```
Comment (tomayac) [inline src/fs-access/file-save.mjs:70]:
```suggestion
  // Use streaming on the `Blob` if the browser supports it.
```

File: src/fs-access/file-save.mjs [hunk 4]
```diff
@@ -67,6 +67,17 @@ export default async (
...
+  if ('stream' in blobOrStream) {
+    const stream = blobOrStream.stream();
+    await stream.pipeTo(writable);
+    return handle;
+    // handle passed readable stream
```
Comment (tomayac) [inline src/fs-access/file-save.mjs:75]:
```suggestion
    // Handle passed `ReadableStream`.
```

File: src/fs-access/file-save.mjs [hunk 5]
```diff
@@ -67,6 +67,17 @@ export default async (
...
+  } else if ('readable' in blobOrStream) {
+    await blobOrStream.readable.pipeTo(writable);
+    return handle;
+  }
+  // default case of blob passed and streams not supported
```
Comment (tomayac) [inline src/fs-access/file-save.mjs:80]:
```suggestion
  // Default case of `Blob` passed and `Blob.stream()` not supported.
```

File: src/legacy/file-save.mjs [hunk 6]
```diff
@@ -19,13 +19,18 @@
...
  * @type { typeof import("../../index").fileSave }
  */
-export default async (blob, options = {}) => {
+export default async (blobOrStream, options = {}) => {
   if (Array.isArray(options)) {
     options = options[0];
   }
   const a = document.createElement('a');
+  let data = blobOrStream;
+  // handle the case where input is a stream
```
Comment (tomayac) [inline src/legacy/file-save.mjs:28]:
```suggestion
  // Handle the case where input is a `ReadableStream`.
```

File: src/legacy/file-save.mjs [hunk 7]
```diff
@@ -19,13 +19,18 @@
...
+  if ("readable" in blobOrStream) {
```
Comment (tomayac) [inline src/legacy/file-save.mjs:29]: Please run `npm run fix` to use the same quote style everywhere.

File: src/legacy/file-save.mjs [hunk 8]
```diff
@@ -44,3 +49,38 @@ export default async (blob, options = {}) => {
   });
   a.click();
 };
+
+/**
+ * Converts a passed readable stream to a blob
```
Comment (tomayac) [inline src/legacy/file-save.mjs:54]:
```suggestion
 * Converts a passed `ReadableStream` to a `Blob`.
```

File: src/legacy/file-save.mjs [hunk 9]
```diff
@@ -44,3 +49,38 @@ export default async (blob, options = {}) => {
...
+ * @param {string} type
+ * @returns {Promise<Blob>}
+ */
+async function streamToBlob(stream, type) {
+  const reader = stream.getReader();
+  const pumpedStream = new ReadableStream({
+    start(controller) {
+      return pump();
+      /**
+       * recursively pumps data chunks out of the stream
```
Comment (tomayac) [inline src/legacy/file-save.mjs:65]:
```suggestion
       * Recursively pumps data chunks out of the `ReadableStream`.
```

File: src/legacy/file-save.mjs [hunk 10]
```diff
@@ -44,3 +49,38 @@ export default async (blob, options = {}) => {
...
+       * @type { () => Promise<void> }
+       */
+      async function pump() {
+        return reader.read().then(({ done, value }) => {
+          // When no more data needs to be consumed, close the stream
```
Comment (tomayac) [inline src/legacy/file-save.mjs:70]:
```suggestion
```

File: src/legacy/file-save.mjs [hunk 11]
```diff
@@ -44,3 +49,38 @@ export default async (blob, options = {}) => {
...
+          if (done) {
+            controller.close();
+            return;
+          }
+          // Enqueue the next data chunk into our target stream
```
Comment (tomayac) [inline src/legacy/file-save.mjs:75]:
```suggestion
```

**Review by tomayac — CHANGES_REQUESTED** [review]
Some additional comment suggestions.

File: index.d.ts (hunk 1 above)
Comment (seanaye) [reply index.d.ts:102]: @tomayac Yes the mime type is required here, there is no easy way to detect the mimeType of the data coming through the stream without adding new dependencies. The mime type is required to build the blob in legacy mode

**Comment by seanaye** [comment]
@googlebot I signed it!

File: index.d.ts (hunk 1 above)
Comment (tomayac) [inline index.d.ts:102]:
```suggestion
  /** To-be-saved `Blob` or `Response` */
  blobOrResponse: Blob | Response,
```

File: src/fs-access/file-save.mjs [hunk 12]
```diff
@@ -20,7 +20,7 @@
  * @type { typeof import("../../index").fileSave }
  */
 export default async (
-  blob,
+  blobOrStream,
```
Comment (tomayac) [inline src/fs-access/file-save.mjs:23]:
```suggestion
  blobOrResponse,
```

File: src/fs-access/file-save.mjs [hunk 13]
```diff
@@ -36,14 +36,14 @@ export default async (
       accept: {},
     };
     if (option.mimeTypes) {
-      if (i === 0) {
-        option.mimeTypes.push(blob.type);
+      if (i === 0 && blobOrStream.type) {
+        option.mimeTypes.push(blobOrStream.type);
       }
```
Comment (tomayac) [inline src/fs-access/file-save.mjs:41]:
```suggestion
      if (i === 0) {
        if (blobOrResponse.type) {
          option.mimeTypes.push(blobOrResponse.type);
        } else if (blobOrResponse.headers && blobOrResponse.headers.get('content-type')) {
          option.mimeTypes.push(blobOrResponse.headers.get('content-type'));
        }
      }
```

File: src/fs-access/file-save.mjs [hunk 14]
```diff
@@ -36,14 +36,14 @@ export default async (
...
       option.mimeTypes.map((mimeType) => {
         types[i].accept[mimeType] = option.extensions || [];
       });
-    } else {
-      types[i].accept[blob.type] = option.extensions || [];
+    } else if (blobOrStream.type) {
+      types[i].accept[blobOrStream.type] = option.extensions || [];
```
Comment (tomayac) [inline src/fs-access/file-save.mjs:46]:
```suggestion
    } else if (blobOrResponse.type) {
      types[i].accept[blobOrResponse.type] = option.extensions || [];
```

File: src/fs-access/file-save.mjs [hunk 15]
```diff
@@ -67,6 +67,17 @@ export default async (
       excludeAcceptAllOption: options[0].excludeAcceptAllOption || false,
     }));
   const writable = await handle.createWritable();
+  // Use streaming on the `Blob` if the browser supports it.
+  if ('stream' in blobOrStream) {
```
Comment (tomayac) [inline src/fs-access/file-save.mjs:71]:
```suggestion
  if ('stream' in blobOrResponse) {
```

File: src/fs-access/file-save.mjs [hunk 16]
```diff
@@ -67,6 +67,17 @@ export default async (
...
+    const stream = blobOrStream.stream();
```
Comment (tomayac) [inline src/fs-access/file-save.mjs:72]:
```suggestion
    const stream = blobOrResponse.stream();
```

File: src/fs-access/file-save.mjs [hunk 17]
```diff
@@ -67,6 +67,17 @@ export default async (
...
+    await stream.pipeTo(writable);
+    return handle;
+    // Handle passed `ReadableStream`.
+  } else if ('readable' in blobOrStream) {
```
Comment (tomayac) [inline src/fs-access/file-save.mjs:76]:
```suggestion
  } else if ('body' in blobOrResponse) {
```

File: src/fs-access/file-save.mjs [hunk 18]
```diff
@@ -67,6 +67,17 @@ export default async (
...
+    await blobOrStream.readable.pipeTo(writable);
```
Comment (tomayac) [inline src/fs-access/file-save.mjs:77]:
```suggestion
    await blobOrResponse.body.pipeTo(writable);
```

File: src/legacy/file-save.mjs [hunk 19]
```diff
@@ -19,13 +19,18 @@
  * Saves a file to disk using the legacy `<a download>` method.
  * @type { typeof import("../../index").fileSave }
  */
-export default async (blob, options = {}) => {
+export default async (blobOrStream, options = {}) => {
```
Comment (tomayac) [inline src/legacy/file-save.mjs:22]:
```suggestion
export default async (blobOrResponse, options = {}) => {
```

File: src/legacy/file-save.mjs [hunk 20]
```diff
@@ -19,13 +19,18 @@
...
   if (Array.isArray(options)) {
     options = options[0];
   }
   const a = document.createElement('a');
+  let data = blobOrStream;
```
Comment (tomayac) [inline src/legacy/file-save.mjs:27]:
```suggestion
  let data = blobOrResponse;
```

File: src/legacy/file-save.mjs [hunk 21]
```diff
@@ -19,13 +19,18 @@
...
+  // Handle the case where input is a `ReadableStream`.
+  if ('readable' in blobOrStream) {
```
Comment (tomayac) [inline src/legacy/file-save.mjs:29]:
```suggestion
  if ('body' in blobOrResponse) {
```

File: src/legacy/file-save.mjs [hunk 22]
```diff
@@ -19,13 +19,18 @@
...
+    data = await streamToBlob(blobOrStream.readable, blobOrStream.type);
```
Comment (tomayac) [inline src/legacy/file-save.mjs:30]:
```suggestion
    data = await streamToBlob(blobOrResponse.body, blobOrResponse.headers.get('content-type'));
```

**Review by tomayac — CHANGES_REQUESTED** [review]
OK, here's the way this could work:

You pass it either a `Blob`, or a `Response`, which can contain a `ReadableStream` and has a `type` attribute:

```js
const response = new Response(new ReadableStream(), {
  headers: {
    "content-type": "text/plain",
  },
});
```

This means we don't need the weird object! I have made suggestions to implement this below, but I probably missed stuff (did it on the front-end without actually testing), so please carefully check this again.

**Comment by tomayac** [comment]
@seanaye Are you OK with the solution I sketched in https://github.com/GoogleChromeLabs/browser-fs-access/pull/67#pullrequestreview-750160960?

**Comment by seanaye** [comment]
Yea definitely, sorry haven't had the time to make the changes

**Review by tomayac — APPROVED** [review]
LGTM

**Pull request merged**